ink_linting/src/mapping_initialized.rs

```diff
@@ -0,0 +1,345 @@
+        Some(InkAttribute::Constructor)
+    } else {
+        None
+    }
+}
+
+impl<'tcx> LateLintPass<'tcx> for MappingInitialized {
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'_>) {
+        let attrs = cx.tcx.hir().attrs(item.hir_id());
+        let ink_attrs = get_ink_attribute(attrs);
+        if ink_attrs.is_none() {
+            return;
+        }
+
+        if let ItemKind::Struct(variant_data, _) = &item.kind {
+            if let VariantData::Unit(..) = variant_data {
+                return;
+            }
+            check_struct(cx, item, variant_data);
+        }
+    }
+}
+
+/// Examines a `struct`. If the struct is annotated as an ink! storage struct
+/// we examine if there is a `ink_storage::Mapping` in it; in case there
+/// is we continue checking all ink! constructors for correct usage of the
+/// `ink_lang::utils::initialize_contract` API.
+///
+/// This function is backwards compatible, in case no annotated ink! struct is
+/// found nothing happens.
+fn check_struct<'a>(cx: &LateContext<'a>, item: &'a Item, data: &VariantData) {
+    let attrs = cx.tcx.hir().attrs(item.hir_id());
+    match get_ink_attribute(attrs) {
+        Some(InkAttribute::Storage) => {}
+        _ => return,
+    }
+
+    let mut marker = None;
+    let fields = data.fields();
+    let storage_contains_mapping = fields.iter().any(|field| {
+        let re = Regex::new(r"ink_storage\[.*\]::lazy::mapping::Mapping")
+            .expect("failed creating regex");
+        if re.is_match(&format!("{:?}", field.ty.kind)) {
+            marker = Some(field);
+            return true;
+        }
+        false
+    });
+
+    if !storage_contains_mapping {
+        log::debug!("Found `#[ink(storage)]` struct without `Mapping`");
+        return;
+    }
+    log::debug!("Found `#[ink(storage)]` struct with `Mapping`");
+
+    let inherent_impls = cx.tcx.inherent_impls(item.def_id);
+    let constructors_without_initialize: Vec<Span> = inherent_impls
+        .iter()
+        .map(|imp_did| item_from_def_id(cx, *imp_did))
+        .filter_map(|item| constructor_but_no_initialize(cx, item))
+        .collect();
+    log::debug!(
+        "Result of searching for constructors without initialize: {:?}",
+        constructors_without_initialize
+    );
+
+    if_chain! {
+        if storage_contains_mapping;
+        if !constructors_without_initialize.is_empty();
+        then {
+            let constructor_span = constructors_without_initialize.get(0)
+                .expect("at least one faulty constructor must exist");
+            let snippet = snippet_opt(cx, *constructor_span)
+                .expect("snippet must exist");
+            span_lint_and_then(
+                cx,
+                MAPPING_INITIALIZED,
+                item.span,
+                &format!(
+                    "you have declared an `#[ink(storage)]` on `{}` without initializing it in the contract constructor.",
```

Comment:
Needs to be more specific, and perhaps without naming and shaming the user. If you look at the Rust compiler errors they are normally in the passive voice https://doc.rust-lang.org/error-index.html.

Here's a few ideas (though try and come up with something more readable!)

"missing contract initialization in constructor for #[ink(storage)] struct with a Mapping field."
"Storage struct with a Mapping field must be initialized on construction"